src/Data/BTree/Primitives.hs

```diff
@@ -43,7 +52,26 @@ data Node height key val where
             } -> Node ('S height) key val
     Leaf :: { leafItems        ::  Map key val
             } -> Node 'Z key val
-    deriving Typeable
+    deriving (Typeable)
+
+instance (Eq key, Eq val) => Eq (Node 'Z key val) where
+    Leaf x == Leaf y = x == y
+
+instance (Eq key, Eq val) => Eq (Node ('S h) key val) where
+    Idx x == Idx y = x == y
+
+data BNode = BIdx | BLeaf deriving Generic
+instance Binary BNode where
+
+instance (Binary key, Binary val) => Binary (Node 'Z key val) where
+    put (Leaf items) = put BLeaf >> put items
+    get = get >>= \case BLeaf -> Leaf <$> get
```

Comment:
Haha nice. The `fail`s suggests that you encode too much, because the information seems to be already available!? So why not remove it? I would still suggest leaving as is, because like this you can simply parse an arbitrary page. Without the `BNode` information you can only parse pages in a context.